### PR TITLE
feature: make API more convenient

### PR DESCRIPTION
**Pull request by fortuna**
- Consolidate the shadowsocks package. No more client subpackage. This makes all the API available in one place.
- The consolidation allows me to remove "Shadowsocks" from the name of functions and types.
- Replace `net.UDPAddr` and `net.TCPAddr` with strings, which support domain names and are easier to work with.
- Add convenience `TCPStreamDialer` to convert a Dialer to a StreamDialer.
- Hide `Cipher` type and make `NewEncryptionKey` take a string which is easier to use
- Add tests.

This makes the library easier and more pleasant to use. I think I'm finally happy with how the transport API is looking like.

### CODE REVIEW

**Comment by fortuna** [comment]
@jyyi1 Thoughts?

File: transport/shadowsocks/cipher.go [hunk 1]
```diff
@@ -26,8 +26,7 @@ import (
 	"golang.org/x/crypto/hkdf"
 )
 
-type Cipher struct {
```
Comment (jyyi1) [inline transport/shadowsocks/cipher.go:29]: Maybe I was wrong, but I remembered `Cipher` is already referenced in tun2socks right?

File: transport/shadowsocks/salt_test.go [hunk 2]
```diff
@@ -42,3 +42,57 @@ func BenchmarkRandomSaltGenerator(b *testing.B) {
...
+
+	for _, b := range output[12:] {
+		if b != 0xFF {
+			t.Error("unexpected zero bit")
+		}
+	}
+}
+
+// Test that all bytes are random when the prefix is nil
+func TestNilPrefix(t *testing.T) {
```
Comment (jyyi1) [inline transport/shadowsocks/salt_test.go:86]: Maybe add test case for empty array in this function as well?

```suggestion
func TestNilOrEmptyPrefix(t *testing.T) {
```

File: transport/shadowsocks/stream.go [hunk 3]
```diff
@@ -63,9 +63,9 @@ var (
 	_ io.ReaderFrom = (*Writer)(nil)
 )
 
-// NewShadowsocksWriter creates a [Writer] that encrypts the given [io.Writer] using
+// NewWriter creates a [Writer] that encrypts the given [io.Writer] using
 // the shadowsocks protocol with the given encryption key.
-func NewShadowsocksWriter(writer io.Writer, key *EncryptionKey) *Writer {
+func NewWriter(writer io.Writer, key *EncryptionKey) *Writer {
```
Comment (jyyi1) [inline transport/shadowsocks/stream.go:68]: The `Writer` type is inconsistent with the following `Reader` type, one is a `struct` while the other is an `interface`. Since both are exported types, I think this might confuse the caller.

**Review by jyyi1 — APPROVED** [review]
The structure looks good, I left a few comments.

File: transport/shadowsocks/cipher.go (hunk 1 above)
Comment (fortuna) [reply transport/shadowsocks/cipher.go:29]: Yes, this is a breaking change. tun2socks is referring to the current version, so it doesn't break. I need to submit this PR first before I can update the dependency there.

File: transport/shadowsocks/stream.go (hunk 3 above)
Comment (fortuna) [reply transport/shadowsocks/stream.go:68]: Good point! I'll add that to my list of things to refine.

File: transport/shadowsocks/salt_test.go (hunk 2 above)
Comment (fortuna) [reply transport/shadowsocks/salt_test.go:86]: We are passing `nil` in the test, not an empty `[]byte`

File: transport/shadowsocks/cipher.go (hunk 1 above)
Comment (fortuna) [reply transport/shadowsocks/cipher.go:29]: Update at https://github.com/Jigsaw-Code/outline-ss-server/pull/152
